src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/DefaultShapeTableManager.cs

```diff
@@ -19,109 +19,119 @@ namespace OrchardCore.DisplayManagement.Descriptors
+            HashSet<string> excludedFeatures;
 
-                    BuildDescriptors(bindingStrategy, builtAlterations, shapeDescriptors);
-                }
+            // Here we don't use a lock for thread safety but for atomicity.
+            lock (_syncLock)
+            {
+                excludedFeatures = new HashSet<string>(_shapeDescriptors.Select(kv => kv.Value.Feature.Id));
+            }
 
-                // Here we don't use a lock for thread safety but for atomicity.
-                lock (_syncLock)
-                {
-                    foreach (var kv in shapeDescriptors)
-                    {
-                        _shapeDescriptors[kv.Key] = kv.Value;
-                    }
-                }
+            var shapeDescriptors = new Dictionary<string, FeatureShapeDescriptor>();
 
-                var enabledAndOrderedFeatureIds = (await _shellFeaturesManager.GetEnabledFeaturesAsync())
-                    .Select(f => f.Id)
-                    .ToList();
+            foreach (var bindingStrategy in bindingStrategies)
+            {
+                var strategyFeature = typeFeatureProvider.GetFeatureForDependency(bindingStrategy.GetType());
+
+                var builder = new ShapeTableBuilder(strategyFeature, excludedFeatures);
+                await bindingStrategy.DiscoverAsync(builder);
+                var builtAlterations = builder.BuildAlterations();
+
+                BuildDescriptors(bindingStrategy, builtAlterations, shapeDescriptors);
+            }
 
-                // let the application acting as a super theme for shapes rendering.
-                if (enabledAndOrderedFeatureIds.Remove(_hostingEnvironment.ApplicationName))
+            // Here we don't use a lock for thread safety but for atomicity.
+            lock (_syncLock)
+            {
+                foreach (var kv in shapeDescriptors)
                 {
-                    enabledAndOrderedFeatureIds.Add(_hostingEnvironment.ApplicationName);
+                    _shapeDescriptors[kv.Key] = kv.Value;
                 }
+            }
 
-                var descriptors = _shapeDescriptors
-                    .Where(sd => enabledAndOrderedFeatureIds.Contains(sd.Value.Feature.Id))
-                    .Where(sd => IsModuleOrRequestedTheme(sd.Value.Feature, themeId))
-                    .OrderBy(sd => enabledAndOrderedFeatureIds.IndexOf(sd.Value.Feature.Id))
-                    .GroupBy(sd => sd.Value.ShapeType, StringComparer.OrdinalIgnoreCase)
-                    .Select(group => new ShapeDescriptorIndex
-                    (
-                        shapeType: group.Key,
-                        alterationKeys: group.Select(kv => kv.Key),
-                        descriptors: _shapeDescriptors
-                    ))
-                    .ToList();
+            var enabledAndOrderedFeatureIds = (await shellFeaturesManager.GetEnabledFeaturesAsync())
+                .Select(f => f.Id)
```

Comment:
```suggestion
                .OrderBy(f => f.Id)
```

Reply:
Why change the previous code?

Reply:
I'm not sure how these Ids are ordered without `OrderBy`?